File: src-tauri/src/ai_runtime/providers/kie.rs
use crate::ai_runtime::errors::{AiResult, AiRuntimeError};
use crate::ai_runtime::polling::{cancelled_error, timeout_error, wait_interval_ms};
use crate::ai_runtime::providers::ProviderExecutionInput;
use crate::ai_runtime::task_registry;
use crate::ai_runtime::types::{GenerateStatus, ProviderExecutionResult};
use serde_json::Value;

const KIE_BASE_URL: &str = "https://api.kie.ai";
const KIE_STATUS_ENDPOINT: &str = "/api/v1/jobs/recordInfo";

pub async fn execute(input: ProviderExecutionInput<'_>) -> AiResult<ProviderExecutionResult> {
    let endpoint = normalize_endpoint(KIE_BASE_URL, input.route);
    let response = send_create_task(&input, &endpoint).await?;

    let final_payload = if let Some(task_id) = extract_task_id(&response) {
        poll_task(&input, &task_id).await?
    } else {
        response
    };

    let urls = extract_urls(&final_payload);
    if urls.is_empty() {
        return Err(AiRuntimeError::new(
            "empty_result",
            "KIE response has no media URL",
        ));
    }

    Ok(ProviderExecutionResult {
        status: GenerateStatus::Completed,
        url: urls.join("|||"),
        metadata: final_payload,
    })
}

async fn send_create_task(input: &ProviderExecutionInput<'_>, endpoint: &str) -> AiResult<Value> {
    let response = input
        .client
        .post(endpoint)
        .bearer_auth(input.api_key)
        .header("Content-Type", "application/json")
        .json(input.body)
        .send()
        .await?;

    let status = response.status();
    let payload = response
        .json::<Value>()
        .await
        .map_err(|e| AiRuntimeError::new("invalid_json", e.to_string()))?;

    if !status.is_success() {
        return Err(AiRuntimeError::new(
            "provider_http_error",
            format!("KIE HTTP {}: {}", status, payload),
        ));
    }

    let code = payload.get("code").and_then(Value::as_i64).unwrap_or(200);
    if code != 200 {
        let message = payload
            .get("msg")
            .and_then(Value::as_str)
            .unwrap_or("KIE create task failed");
        return Err(AiRuntimeError::new("provider_task_failed", message));
    }

    Ok(payload)
}

async fn poll_task(input: &ProviderExecutionInput<'_>, task_id: &str) -> AiResult<Value> {
    let interval = input.polling.map(|p| p.interval).unwrap_or(3000);
    let max_attempts = input.polling.map(|p| p.max_attempts).unwrap_or(200);

    for _ in 0..max_attempts {
        if task_registry::is_cancelled(input.request_id) {
            return Err(cancelled_error(input.request_id));
        }

        wait_interval_ms(interval).await;

        let endpoint = format!("{}{}?taskId={}", KIE_BASE_URL, KIE_STATUS_ENDPOINT, task_id);
        let response = input
            .client
            .get(&endpoint)
            .bearer_auth(input.api_key)
            .send()
            .await?;

        let payload = response
            .json::<Value>()
            .await
            .map_err(|e| AiRuntimeError::new("invalid_json", e.to_string()))?;

        let code = payload.get("code").and_then(Value::as_i64).unwrap_or(200);
        if code != 200 {
            continue;
        }

        let state = payload
            .pointer("/data/state")
            .and_then(Value::as_str)
            .unwrap_or("");

        if state == "success" {
            return Ok(payload);
        }

        if state == "fail" {
            let fail_msg = payload
                .pointer("/data/failMsg")
                .and_then(Value::as_str)
                .unwrap_or("KIE task failed");
            return Err(AiRuntimeError::new("provider_task_failed", fail_msg));
        }
    }

    Err(timeout_error(max_attempts))
}

fn extract_task_id(payload: &Value) -> Option<String> {
    payload
        .pointer("/data/taskId")
        .and_then(Value::as_str)
        .map(ToString::to_string)
}

fn extract_urls(payload: &Value) -> Vec<String> {
    if let Some(result_json) = payload
        .pointer("/data/resultJson")
        .and_then(Value::as_str)
    {
        if let Ok(parsed) = serde_json::from_str::<Value>(result_json) {
            let mut urls = vec![];
            extract_from_array(parsed.get("resultUrls"), &mut urls);
            extract_from_array(parsed.get("images"), &mut urls);
            extract_from_array(parsed.get("videos"), &mut urls);
            if !urls.is_empty() {
                return urls;
            }
        }
    }

    let mut urls = vec![];
    extract_from_array(payload.get("resultUrls"), &mut urls);
    extract_from_array(payload.get("images"), &mut urls);
    extract_from_array(payload.get("videos"), &mut urls);
    urls
}

fn extract_from_array(value: Option<&Value>, target: &mut Vec<String>) {
    let Some(Value::Array(items)) = value else {
        return;
    };

    for item in items {
        if let Some(url) = item.as_str() {
            target.push(url.to_string());
            continue;
        }

        if let Some(url) = item.get("url").and_then(Value::as_str) {
            target.push(url.to_string());
        }
    }
}

fn normalize_endpoint(base: &str, route: &str) -> String {
    if route.starts_with("http://") || route.starts_with("https://") {
        return route.to_string();
    }
    if route.starts_with('/') {
        return format!("{}{}", base, route);
    }
    format!("{}/{}", base, route)
}

File: electron/main/services/ai-runtime/providers/kie.ts
import { AiRuntimeError } from '../errors'
import { POLL_QUERY_FAILED, pollUntilResult } from '../polling'
import type { JsonValue, ProviderContinuePollingInput, ProviderExecutionInput, ProviderExecutionResult } from '../types'
import { getPointer, normalizeEndpoint, pushUniqueUrl, readJsonResponse, stringAt } from './helpers'

const KIE_BASE_URL = 'https://api.kie.ai'
const KIE_STATUS_ENDPOINT = '/api/v1/jobs/recordInfo'

export async function execute(input: ProviderExecutionInput): Promise<ProviderExecutionResult> {
  const response = await sendCreateTask(input, normalizeEndpoint(KIE_BASE_URL, input.route))
  const taskId = extractTaskId(response)
  if (taskId) {
    return { status: 'pending', url: '', taskId, metadata: response }
  }

  const urls = extractUrls(response)
  if (urls.length === 0) {
    throw new AiRuntimeError('empty_result', `KIE response has no media URL (task_id=${taskId ?? 'unknown'})`)
  }
  return { status: 'completed', url: urls.join('|||'), taskId, metadata: response }
}

export async function continuePolling(input: ProviderContinuePollingInput): Promise<ProviderExecutionResult> {
  const finalPayload = await pollTask(input)
  const urls = extractUrls(finalPayload)
  if (urls.length === 0) {
    throw new AiRuntimeError('empty_result', `KIE response has no media URL (task_id=${input.taskId})`)
  }
  return { status: 'completed', url: urls.join('|||'), taskId: input.taskId, metadata: finalPayload }
}

async function sendCreateTask(input: ProviderExecutionInput, endpoint: string): Promise<JsonValue> {
  const response = await fetch(endpoint, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${input.apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(input.body),
    signal: input.signal,
  })
  const payload = await readJsonResponse(response, 'KIE')
  const code = getPointer(payload, '/code')
  if (typeof code === 'number' && code !== 200) {
    throw new AiRuntimeError('provider_task_failed', stringAt(payload, '/msg') ?? 'KIE create task failed')
  }
  return payload
}

async function pollTask(input: ProviderContinuePollingInput): Promise<JsonValue> {
  return await pollUntilResult(input, async () => {
    const endpoint = `${KIE_BASE_URL}${KIE_STATUS_ENDPOINT}?taskId=${encodeURIComponent(input.taskId)}`
    const response = await fetch(endpoint, {
      headers: { Authorization: `Bearer ${input.apiKey}` },
      signal: input.signal,
    })
    const payload = await readJsonResponse(response, 'KIE')
    const code = getPointer(payload, '/code')
    if (typeof code === 'number' && code !== 200) {
      // 查询本身没成功（含 taskId 已被清理的情况），不是"任务还在跑"
      return POLL_QUERY_FAILED
    }
    const state = stringAt(payload, '/data/state') ?? ''
    if (state === 'success') return payload
    if (state === 'fail') {
      throw new AiRuntimeError('provider_task_failed', stringAt(payload, '/data/failMsg') ?? 'KIE task failed')
    }
    return undefined
  })
}

function extractTaskId(payload: JsonValue): string | undefined {
  return stringAt(payload, '/data/taskId')
}

function extractUrls(payload: JsonValue): string[] {
  const resultJson = stringAt(payload, '/data/resultJson')
  if (resultJson) {
    try {
      const parsed = JSON.parse(resultJson) as JsonValue
      const urls: string[] = []
      extractFromArray(getPointer(parsed, '/resultUrls'), urls)
      extractFromArray(getPointer(parsed, '/images'), urls)
      extractFromArray(getPointer(parsed, '/videos'), urls)
      if (urls.length > 0) return urls
    } catch {
      // Fall through to direct payload extraction.
    }
  }
  const urls: string[] = []
  extractFromArray(getPointer(payload, '/resultUrls'), urls)
  extractFromArray(getPointer(payload, '/images'), urls)
  extractFromArray(getPointer(payload, '/videos'), urls)
  return urls
}

function extractFromArray(value: JsonValue | undefined, target: string[]): void {
  if (!Array.isArray(value)) return
  for (const item of value) {
    if (typeof item === 'string') pushUniqueUrl(target, item)
    else {
      const url = getPointer(item, '/url')
      if (typeof url === 'string') pushUniqueUrl(target, url)
    }
  }
}
